pkg/svcat/service-catalog/binding_test.go

```diff
@@ -0,0 +1,213 @@
+			Expect(err).To(HaveOccurred())
+			Expect(err.Error()).Should(ContainSubstring(errorMessage))
+			Expect(badClient.Actions()[0].Matches("list", "servicebindings")).To(BeTrue())
+		})
+	})
+
+	Describe("Bind", func() {
+		It("Calls the generated v1beta1 method to create a binding", func() {
+			bindingNamespace := "banana_namespace"
+			bindingName := "banana_binding"
+			instanceName := "banana_instance"
+			binding, err := sdk.Bind(bindingNamespace, bindingName, instanceName, "banana_secret", map[string]string{}, map[string]string{})
+
+			Expect(err).NotTo(HaveOccurred())
+			Expect(binding).NotTo(BeNil())
+			Expect(binding.ObjectMeta.Namespace).To(Equal(bindingNamespace))
+			Expect(binding.ObjectMeta.Name).To(Equal(bindingName))
+			Expect(binding.Spec.ServiceInstanceRef.Name).To(Equal(instanceName))
+			Expect(svcCatClient.Actions()[0].Matches("create", "servicebindings")).To(BeTrue())
+		})
+
+		It("Bubbles up errrors", func() {
+			badClient := &fake.Clientset{}
+			errorMessage := "error retrieving list"
+			badClient.AddReactor("create", "servicebindings", func(action testing.Action) (bool, runtime.Object, error) {
+				return true, nil, fmt.Errorf(errorMessage)
+			})
+			sdk.ServiceCatalogClient = badClient
+
+			bindingNamespace := "banana_namespace"
+			bindingName := "banana_binding"
+			instanceName := "banana_instance"
+			binding, err := sdk.Bind(bindingNamespace, bindingName, instanceName, "banana_secret", map[string]string{}, map[string]string{})
+
+			Expect(binding).To(BeNil())
+			Expect(err).To(HaveOccurred())
+			Expect(err.Error()).Should(ContainSubstring(errorMessage))
+			Expect(badClient.Actions()[0].Matches("create", "servicebindings")).To(BeTrue())
+		})
+	})
+
+	Describe("Unbind", func() {
+		It("Calls the generated v1beta1 method to delete a binding", func() {
+			instanceNamespace := sb.Namespace
+			instanceName := "apple_instance"
+			si := &v1beta1.ServiceInstance{ObjectMeta: metav1.ObjectMeta{Name: instanceName, Namespace: instanceNamespace}}
+			sb.Spec.ServiceInstanceRef.Name = si.Name
+			svcCatClient = fake.NewSimpleClientset(sb, sb2)
+			sdk = &SDK{
+				ServiceCatalogClient: svcCatClient,
+			}
+
+			err := sdk.Unbind(instanceNamespace, instanceName)
+
+			Expect(err).NotTo(HaveOccurred())
+			Expect(svcCatClient.Actions()[0].Matches("list", "servicebindings")).To(BeTrue())
+			Expect(svcCatClient.Actions()[1].Matches("delete", "servicebindings")).To(BeTrue())
+		})
+
+		It("Bubbles up errrors", func() {
+			instanceNamespace := sb.Namespace
+			instanceName := "apple_instance"
+			errorMessage := "error deleting binding"
+			si := &v1beta1.ServiceInstance{ObjectMeta: metav1.ObjectMeta{Name: instanceName, Namespace: instanceNamespace}}
+			sb.Spec.ServiceInstanceRef.Name = si.Name
+			//svcCatClient = fake.NewSimpleClientset(sb, sb2)
+			svcCatClient := &fake.Clientset{}
+			svcCatClient.AddReactor("list", "servicebindings", func(action testing.Action) (bool, runtime.Object, error) {
```

Comment:
This is just a question for my own understanding. For the failure mode test cases, would the test also work if we didn't reset `svcCatClient` to a new fake clientset, and then still setup a reactor for the call that we want to sabotage.

Just wondering if we can remove some of the extra setup for the failure cases or if that's not how the fake client works.

Reply:
Same reason as up above. They're passing in by value not reference, so I have to recreate the SDK to modify the fake client. We could change the SDK to accept a pointer to a client instead of an actual client to make it work like that, though. Hmm.